core/common/src/LocalDateRange.kt

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2019-2022 JetBrains s.r.o.
+ * Use of this source code is governed by the Apache 2.0 License that can be found in the LICENSE.txt file.
+ */
+
+package kotlinx.datetime
+
+import kotlin.random.Random
+import kotlin.random.nextInt
+
+internal class LocalDateProgressionIterator(private val iterator: IntIterator) : Iterator<LocalDate> {
+    override fun hasNext(): Boolean = iterator.hasNext()
+    override fun next(): LocalDate = LocalDate.fromEpochDays(iterator.next())
+}
+
+public open class LocalDateProgression
+internal constructor(internal val intProgression: IntProgression) : Iterable<LocalDate> {
+
+    internal constructor(
+        start: LocalDate,
+        endInclusive: LocalDate,
+        step: Int
+    ) : this(IntProgression.fromClosedRange(start.toEpochDays(), endInclusive.toEpochDays(), step))
+
+    public val first: LocalDate = LocalDate.fromEpochDays(intProgression.first)
+    public val last: LocalDate = LocalDate.fromEpochDays(intProgression.last)
+
+    override fun iterator(): Iterator<LocalDate> = LocalDateProgressionIterator(intProgression.iterator())
+
+    public open fun isEmpty(): Boolean = intProgression.isEmpty()
+
+    override fun toString(): String = if (intProgression.step > 0) "$first..$last step ${intProgression.step}" else "$first downTo $last step ${intProgression.step}"
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other == null || this::class != other::class) return false
+
+        other as LocalDateProgression
+
+        return intProgression == other.intProgression
+    }
+
+    override fun hashCode(): Int {
+        return intProgression.hashCode()
+    }
+
+
+    public companion object {
+        public fun fromClosedRange(
+            rangeStart: LocalDate,
+            rangeEnd: LocalDate,
+            stepValue: Int,
+            stepUnit: DateTimeUnit.DayBased
+        ): LocalDateProgression = LocalDateProgression(rangeStart, rangeEnd, stepValue * stepUnit.days)
+    }
+}
+
+public class LocalDateRange(start: LocalDate, endInclusive: LocalDate) : LocalDateProgression(start, endInclusive, 1), ClosedRange<LocalDate>, OpenEndRange<LocalDate> {
+    override val start: LocalDate get() = first
+    override val endInclusive: LocalDate get() = last
+    override val endExclusive: LocalDate get() = endInclusive.plus(1, DateTimeUnit.DAY)
```

Comment:
See the proposed API: this function should be deprecated, as it can throw an exception on overflow, which may be surprising. Also, this function should include the documentation and throw `IllegalStateException` if the overflow did happen. See https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.ranges/-char-range/end-exclusive.html

Reply:
Will add check and deprecation warning